library/firewall_lib.py

```diff
@@ -964,6 +973,495 @@ def set_target(self, target):
+
+    def change(self, *args):
+        """Like cmd(), but skipped in check_mode.
+
+        Also set self.changed.
+        """
+        if not self.module.check_mode:
+            self.cmd(*args)
+        self.changed = True
+
+    def query(self, *args):
+        """Call firewall-offline-cmd query command, convert exit code to bool."""
+
+        rc = self._call_offline_cmd(args, check_rc=False)[0]
+        return True if rc == 0 else False
+
+    def finalize(self):
+        # nothing to do here, all changes are written immediately in offline mode
+        pass
+
+    def check_state(self, allowed, option):
+        """Check and interpret self.state
+
+        allowed is a list of allowed values. E.g. most operations only accept
+        enabled/disabled, while others only accept present/absent, some accept
+        either. This keeps the behaviour bug-for-bug compatible with
+        OnlineAPIBackend.
+
+        Return True for enabled/present or False for disabled/absent.
+        """
+        if self.state not in allowed:
+            self.module.fail_json(
+                msg="state '%s' not accepted for option '%s'" % (self.state, option)
+            )
+
+        return self.state in ["enabled", "present"]
+
+    def set_firewalld_conf(self, firewalld_conf, allow_zone_drifting_deprecated):
+        if allow_zone_drifting_deprecated:
+            # compatibility with OnlineAPIBackend: allow_zone_drifting gets
+            # ignored when deprecated, without failing the role
+            other_keys = set(firewalld_conf.keys()) - set(["allow_zone_drifting"])
+            if len(other_keys) == 0:
+                # parser in main() already wrote a warning
+                return
+
+        # there are currently no other supported options in firewalld_conf, so
+        # this should not happen; if it ever does, implement it
+        self.module.fail_json(
+            msg="firewalld_conf is not currently supported in offline mode; please file a bug"
+        )
+
+    def set_zone(self):
+        create = self.check_state(["present", "absent"], "zone")
+        if create != self.zone_exists:
+            self.change(
+                "--%s-zone=%s" % ("new" if create else "delete", self.zone),
+            )
+        if not create and self.zone_exists:
+            self.zone = None
+            self.zone_exists = False
+
+    def set_default_zone(self, zone):
+        if self.cmd("--get-default-zone") != zone:
+            self.change("--set-default-zone", zone)
+
+    def set_service(
+        self,
+        service_operation,
+        service,
+        description,
+        short,
+        port,
+        protocol,
+        source_port,
+        helper_module,
+        destination_ipv4,
+        destination_ipv6,
+        includes,
+    ):
+        if service_operation:
+            present = self.check_state(["present", "absent"], "service")
+            known_services = self.cmd("--get-services").split()
```

Comment:
This is ok - firewall allows only alphanum, and some object can have additional characters specified in the `ADDITIONAL_ALNUM_CHARS` for the object type, and none of these are whitespace.  basically `["_", "-", ":", ".", "/"]`